### DISCUSSION TITLE
I trained the model with 2026, but detected 2 same boxes on one object

### DISCUSSION THREAD

**Discussion by fangzhengzhu** [ultralytics/ultralytics#24130] [Q&A]
Hello, when I predicted with the model trained with yolo26, 2 similar boxes  with same label were on one area. According to NMS free and DFL Free, I think there should be only one exists.
![ad123f39-6ed2-47c4-9a32-75f600bc82fb](https://github.com/user-attachments/assets/86305b09-b560-4697-8bc2-e3e9f930f576)

And I also found DFL in the final model file named best.pt, Does it mean the final model is not really a yolo26 model?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fangzhengzhu, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many of the most common questions may already be answered.

Since this appears to be a model behavior/question report, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so the team can better investigate 🔍. Helpful details include:
-Your exact training and prediction commands
-Your `ultralytics` version
-A small sample image or dataset example
-Training logs/results
-Any custom config changes you made

If this is related to custom training ❓, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community 💬.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by fangzhengzhu** [reply]
ultralytics version:  8.4.33

train code:
`from ultralytics import YOLO

# Load a pretrained YOLO26n model
model = YOLO("./model/yolo26m.pt")

# Train the model on the COCO8 dataset for 100 epochs
train_results = model.train(
    data="/home/swzx/source/OCRAutoScore/segmentation/Layout4Card/dataset/dataset.yaml",  # Path to dataset configuration file
    epochs=100,  # Number of training epochs
    imgsz=1024,  # Image size for traininga
    workers=4,
    device=[0,1],  # Device to run on (e.g., 'cpu', 0, [0,1,2,3])
)
`

predict code:
`def predict(self, image_path):
        results = self.model(image_path, conf=0.5, iou=0.6, verbose=False)
        return results
`

partial dataset files:
[label.zip](https://github.com/user-attachments/files/26478296/label.zip)

I didn't save the training logs.

No config changes I had made.

**Reply by glenn-jocher** [reply]
Thanks for the details — with YOLO26 the default end-to-end head does not use NMS, so `iou=0.6` will not change this result; two close boxes usually means the model/data still learned duplicate assignments, and seeing a `DFL` entry in `best.pt` alone does not mean the checkpoint is not YOLO26. Please verify the model is actually running the end-to-end head as shown in the [end-to-end detection guide](https://docs.ultralytics.com/guides/end2end-detection/), and if your labels are segmentation polygons retrain with `yolo26m-seg.pt` instead of `yolo26m.pt`; if it still reproduces, please paste the outputs below and 2-3 raw image/label pairs.

```bash
yolo checks
```

```python
from ultralytics import YOLO

model = YOLO("best.pt")
model.predict("image.jpg", verbose=False)
print(model.predictor.model.end2end)
```

**Comment by Y-T-G** [comment]
How many images do you have for training?

**Reply by Shravasti221** [reply]
Would having less images affect how nms works? What is the expected minimum number of train-time instances of different classes (assuming this is an object detection task) would be required for the model to work as expected? Is there a minimum number of epochs after which the nms stabilizes or is that irrelevant?

**Reply by glenn-jocher** [reply]
Yes — fewer images or low class diversity can make the `end2end` head learn weaker one-to-one assignments, so duplicate boxes may still appear; that’s a data/convergence issue, not “NMS stabilizing.” There’s no hard minimum, but in practice you usually want at least hundreds of labeled instances per class and to train until validation metrics plateau; the [YOLO26 end-to-end detection guide](https://docs.ultralytics.com/guides/end2end-detection) also notes that final detections are produced directly, with NMS only used if `end2end=False` or an export format falls back to the one-to-many head.

**Reply by Shravasti221** [reply]
Thank you Glenn for your response. 
I currently have 147 train images - 3 classes with 92 instances (easy large class), 285 instances and 300 instances each in a very high resolution image. I found that pre-tiling the input image at different sizes (original full frame, 1280, 2560) and then training with imgsz=1280 was required to ensure nms-like functionality was learned by the train set. Otherwise the model was predicting multiple bounding boxes. My question is that: From what I understand the scale parameter handles the zooming in on a portion of the image. If my input dataset has 5k resolution images, will this 
a) first be resized to imgsz=1280 (effectively losing the information) and then scale based augmentations will be applied?
b) or, will it zoom into different sections of the images in the original 5k resolution and then resize to 1280 which will effectively render my pre-tiling redundant? 

I ask this with relation to the above question as the model converges very quickly if I don't pre-tile the input image, but if setting the scale parameter effectively (perhaps more so) what the pre-tiling does, then I want to understand why does my train set converge prematurely when not pre-tiled.

Note: I checked model.predictor.model.end2end and en2end is turned on for my model. And I am not exporting the model to any other format. I am testing the trained model as is.
Also I require the original image resolution (5k) as the bounding box size varies a lot. Around 86% of one class is under 8 px while the other class can exceed 1000 px in width.

**Reply by raimbekovm** [reply]
good question, and it's worth being precise about *where* in the pipeline the resize happens, because that decides the answer. it's your option (a), and your pre-tiling is not redundant - i verified both the pipeline geometry and the effect on the end2end head.

**the resize happens at load time, before any augmentation.** every image goes through `load_image()`, which resizes the long side down to `imgsz` keeping aspect ratio (`ultralytics/data/base.py:240-244`). only then does the aug stack run: `v8_transforms` builds `Mosaic -> RandomPerspective(scale=...) -> ...` (`ultralytics/data/augment.py`), and `scale` is `s = random.uniform(1 - scale, 1 + scale)` applied inside `RandomPerspective` on the already assembled, already downscaled mosaic canvas (`augment.py:1042`). so `scale` is zoom jitter on already downscaled pixels, not a sampler that crops native regions out of your 5k image. it cannot bring back detail discarded at load time. measured on a synthetic 5000x4000 image with a 60x48 object at `imgsz=1280`: load resizes it to 1024x1280 and the object drops to ~15x12 px (about 15x fewer pixels) *before any augmentation*; `scale=0.5` then only jitters it between ~3 and ~23 px around that degraded size. a 1280 tile cropped from the native 5k keeps the object at 60x48 px instead.

**effect on the end2end head, measured.** i trained `yolo26n` (end2end=True confirmed) twice with identical model, epochs, hyp defaults and the same number of train images - the only difference being per-object resolution. branch A: native images fed full-frame so `load_image` downscales ~5x (objects end up ~7 px). branch B: the same content pre-tiled so objects keep native pixels (~34 px).

```
                full-frame downscale   pre-tiled
duplicate rate        0.189              0.083      (extra boxes per GT)
recall                0.913              0.972
mAP50                 0.959              0.973
```

so the degraded-resolution model produces ~2.3x more duplicate boxes on the same end2end head, and it does not go away with training (80 epochs, train box loss was actually *higher* throughout for the downscaled branch, 4.1->1.7 vs 1.4->0.5 - the low-resolution input is harder to localize, not easier). this lines up with what was said earlier in the thread: it's a data/resolution issue, not "nms stabilizing" - there is no separate nms step in the end2end path, and more epochs do not fix a resolution-starved input.

**practical takeaways.** keep tiling, it is the correct approach for 5k imagery with small objects, and it is not made redundant by `scale`. raising `imgsz` helps a bit but is still lossy versus native for 5k, so tiling stays better for the smallest objects. and 147 images with ~92/285/300 instances is on the low side for the end2end head, so more labeled instances per class will further reduce duplicates. relevant reading: https://docs.ultralytics.com/guides/sahi-tiled-inference/ and https://docs.ultralytics.com/guides/model-training-tips/
